feat(python): Update docs for SDK 2.0 (#9182)
getsentry/sentry-docs: docs/platforms/python/migration/1.x-to-2.x.mdx

description: "Learn about migrating from sentry-python 1.x to 2.x"
---

This guide describes the common patterns involved in migrating to version 2.x of the `sentry-python` SDK.

> While the top level API stays the same for the most part, there's a number of deprecations and breaking changes in `2.0`. This guide captures the top changes that will affect most users and the corresponding updates they'll need to make in order to migrate. For the full list of changes, check out the [detailed migration guide in the repository](https://github.com/getsentry/sentry-python/blob/sentry-sdk-2.0/MIGRATION_GUIDE.md).

Reviewer: ```suggestion
While the top-level API stays the same for the most part, there's a number of deprecations and breaking changes in `2.0`. This guide captures the top changes that will affect most users and the corresponding updates they'll need to make in order to migrate. For the full list of changes, check out the [detailed migration guide in the repository](https://github.com/getsentry/sentry-python/blob/sentry-sdk-2.0/MIGRATION_GUIDE.md).
```

## Python Version Support

> Sentry Python SDK `2.0` now only supports Python 3.6 and higher. If you're on Python 2.7 or 3.5 and lower, you'll need to stay on 1.x.

Reviewer: ```suggestion
Sentry Python SDK `2.0` now only supports Python 3.6 and higher. If you're on Python 2.7 or 3.5 and lower, you'll need to stay on `1.x`.
```

## Custom Instrumentation

> Hub-based APIs as well as some scope-based APIs like `configure_scope()` and `push_scope()` are now deprecated as a consequence of [reworking](/platforms/python/enriching-events/scopes/) the way the SDK saves and propagates data internally. Instead, SDK 2.0 introduces the concept of a global, isolation and current scope. In short:

Reviewer: ```suggestion
Hub-based APIs as well as some scope-based APIs (like `configure_scope()` and `push_scope()`) are now deprecated as a consequence of [reworking](/platforms/python/enriching-events/scopes/) the way the SDK saves and propagates data internally. Instead, SDK 2.0 introduces the concept of a global, isolation and current scope. In short:
```

> - The **global scope** holds data that doesn't change as long as your app is running, for instance the current release.

Reviewer: ```suggestion
- The **global scope** holds data that doesn't change as long as your app is running (for example, the current release).
```

> - The **isolation scope** follows a single basic unit of your program such as a single request-response cycle or one task in a task queue and holds data pertaining to it. In most cases, the lifecycle of an isolation scope maps to the lifecycle of a transaction.

Reviewer: ```suggestion
- The **isolation scope** follows a single basic unit of your program, such as a single request-response cycle or one task in a task queue, and holds data pertaining to it. In most cases, the lifecycle of an isolation scope maps to the lifecycle of a transaction.
```

- The **current scope** wraps a smaller part of the code, often something that's being tracked by a single span.

> The SDK will take care of deciding what data belongs on which scope as long as you use the top-level API, so for instance `sentry_sdk.set_tag()` as opposed to `scope.set_tag()`.

Reviewer: ```suggestion
The SDK will take care of deciding what data belongs on which scope as long as you use the top-level API. For example, you can use `sentry_sdk.set_tag()` and let the SDK handle the data for you, instead of manually setting it with `scope.set_tag()`.
```
I thought it could help to explain more for the user here. Hope I interpreted it correctly.

should affect the whole transaction (one request-response cycle, one execution of
a task, and so on) or just a specific part of the code.

If you want your scope change to affect a smaller unit of code such as a span, use the current scope.

```python diff
- with configure_scope() as scope:
-     # do something with scope
+ scope = Scope.get_current_scope()
+ # do something with scope
```

If you want your scope change to affect the whole transaction, use the isolation scope.

```python diff
- with configure_scope() as scope:
-     # do something with scope
+ scope = Scope.get_isolation_scope()
+ # do something with scope
```

Additionally, you can no longer mutate a manually-created transaction in a `configure_scope` block. Instead, you need to get the current scope to mutate the transaction. Here is a recipe on how to change your code to make it work:

```python diff
transaction = sentry_sdk.transaction(...)

# later in the code execution:

- with sentry_sdk.configure_scope() as scope:
-     scope.set_transaction_name("new-transaction-name")

+ scope = sentry_sdk.Scope.get_current_scope()
+ scope.set_transaction_name("new-transaction-name")
```

### Scope Pushing

> Scope pushing translates to forking a scope in `2.0`. Similarly as above, you either want to fork the current or the isolation scope, depending on how long the new scope should be active. If it should encompass the whole transaction, fork the isolation scope, otherwise fork the current scope.

Reviewer: ```suggestion
Scope pushing translates to forking a scope in `2.0`. You either want to fork the current or the isolation scope, depending on how long the new scope should be active. If it should encompass the whole transaction, fork the isolation scope, otherwise fork the current scope.
```
